### PR TITLE
Feature/extend pythonapi with new model.reshape overload

### PR DESCRIPTION
**Pull request by amkarn258**
### Details:
 - *New model.reshape() overload added for list of shapes*

### Tickets:
 - *110444*

Closes Issue [#18347](https://github.com/openvinotoolkit/openvino/issues/18347)

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 1]
```diff
@@ -327,19 +327,44 @@ void regclass_graph_Model(py::module m) {
     model.def(
         "reshape",
         [](ov::Model& self, const py::list& partial_shape) {
-            ov::PartialShape new_shape(Common::partial_shape_from_list(partial_shape));
-            py::gil_scoped_release release;
-            self.reshape(new_shape);
+            if (py::isinstance<py::list>(partial_shape[0])) {
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:330]: what if user put empty list corresponding to scalar? Example: `[]`

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 2]
```diff
@@ -327,19 +327,44 @@ void regclass_graph_Model(py::module m) {
...
+                std::map<ov::Output<ov::Node>, ov::PartialShape> shapes_map;
+
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:332]: should we check a number of shapes match to a number of inputs?

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 1 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:330]: Adding that check and test as well, it's taking a lot of time to build again. will commit once it's done

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:332]: Adding this check too, should we throw an error message if it's not matching? what error message do you suggest?

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 1 above)
Comment (rkazants) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:330]: Please cover this in this PR

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (rkazants) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:332]: Please cover this in this PR

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 1 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:330]: Yes I'll add it in this PR only

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:332]: Yes I'll add it in this PR only

**Comment by amkarn258** [comment]
Hi there, 

Is there a command through which I can only build python api and nothing else? it takes a lot of time to build. I tried searching in the documentation but I couldn't find it.

**Comment by akuporos** [comment]
> Hi there,
> 
> Is there a command through which I can only build python api and nothing else? it takes a lot of time to build. I tried searching in the documentation but I couldn't find it.

Hi,

I guess you may need to turn off building of some OpenVINO components and that will speed up your building.
Try it like this 
`cmake -DENABLE_SAMPLES=OFF -DENABLE_OV_PADDLE_FRONTEND=OFF -DENABLE_OV_TF_FRONTEND=OFF -DENABLE_OV_ONNX_FRONTEND=OFF -DENABLE_TESTS=OFF -DENABLE_INTEL_GPU=OFF -DCMAKE_BUILD_TYPE=Release -DENABLE_WHEEL=OFF ..`

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:332]: @rkazants this check is already there, when the number of shapes don't match number of inputs it throws this error:

`FAILED tests/test_runtime/test_model.py::test_reshape_with_list_of_shapes - AssertionError: assert 4 == 3`

when input shape is 4 and in shapes list there are 3

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 1 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp]: added a check if it's not an empty list

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:332]: May be add such a test with failure?
https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions
And you may use parametrization like [here](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/tests/test_runtime/test_model.py#L47)

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 3]
```diff
@@ -327,19 +327,44 @@ void regclass_graph_Model(py::module m) {
...
         py::arg("partial_shape"),
         R"(
                 Reshape model input.
 
                 GIL is released while running this function.
 
-                :param partial_shape: New shape.
-                :type partial_shape: list
+                :param partial_shape: New shape or list of shapes.
+                :type partial_shape: list or list of lists
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:354]:
```suggestion
                :type partial_shape: Union[List[int], List[List[int]]]
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:335]: @akuporos have added this now

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 4]
```diff
@@ -327,19 +327,47 @@ void regclass_graph_Model(py::module m) {
...
         py::arg("partial_shape"),
         R"(
                 Reshape model input.
 
                 GIL is released while running this function.
 
-                :param partial_shape: New shape.
-                :type partial_shape: list
+                :param partial_shape: New shape or list of shapes.
+                :type partial_shape: Union[List[int], List[List[int]]]
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:357]: Should we allow the following types to represent `PartialShape`?

            (1) `openvino.runtime.PartialShape`
            (2) `list` consisting of dimensions
            (3) `tuple` consisting of dimensions
            (4) `str`, string representation of `openvino.runtime.PartialShape`

Comment (amkarn258) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:357]: didn't get what you mean, do you mean something like this to be also included in this ticket? 

model.reshape([(1,2), (2,3), [4,5]])

File: src/bindings/python/tests/test_runtime/test_model.py [hunk 5]
```diff
@@ -366,6 +366,33 @@ def check_shape(new_shape):
...
+    def generate_model_with_inputs(input_shapes):
+        params = [ops.parameter(shape, dtype=np.float32) for shape in input_shapes]
+        relus = [ops.relu(param) for param in params]
+        return Model(relus, params)
+
+    model = generate_model_with_inputs(input_shapes)
+
+    with expectation as e:
+        model.reshape(new_shapes_list)
+        if not isinstance(e, type(pytest.raises(ValueError))):
```
Comment (akuporos) [inline src/bindings/python/tests/test_runtime/test_model.py:387]: why is it need here?

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 4 above)
Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:357]: @amkarn258 @rkazants means to extend the docstring with all possible list's element types.

Possible types of list elements you may found [here](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/core/common.cpp#L547).
[Example ](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/graph/model.cpp#L419) of such a docstring

Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:357]: @amkarn258 meanwhile I'd be very great if you extend also your test with these types

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 6]
```diff
@@ -327,19 +327,51 @@ void regclass_graph_Model(py::module m) {
     model.def(
         "reshape",
         [](ov::Model& self, const py::list& partial_shape) {
-            ov::PartialShape new_shape(Common::partial_shape_from_list(partial_shape));
-            py::gil_scoped_release release;
-            self.reshape(new_shape);
+            if (py::len(partial_shape) > 0 && (py::isinstance<py::list>(partial_shape[0]) || py::isinstance<ov::PartialShape>(partial_shape[0]) || py::isinstance<py::str>(partial_shape[0]) || py::isinstance<py::tuple>(partial_shape[0]))) {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:330]: > py::isinstance<ov::PartialShape>(partial_shape[0])

It looks excessive as originally it was list of PartialShapes and now looks like else-statement won't be reached.

**Comment by amkarn258** [comment]
Hi, 

Is the number of arguments also supposed to be different now (for this new overload which takes list of shapes)? as model.reshape() method now takes 2 arguments instead of 1. 

@rzubarev @akuporos

**Comment by akuporos** [comment]
@amkarn258 yes, I think it also should be supported and have default value

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 7]
```diff
@@ -353,52 +353,66 @@ void regclass_graph_Model(py::module m) {
...
-                :param partial_shape: New shape.
-                :type partial_shape: openvino.runtime.PartialShape
-                :param variables_shapes: New shapes for variables
-                :type variables_shapes: Dict[keys, values]
-                :return : void
              )");
 
     model.def(
         "reshape",
-        [](ov::Model& self, const py::list& partial_shape, const py::dict& variables_shapes) {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:397]: Could you please switch these overloads back?|
Just to have less diff here

**Comment by mlukasze** [comment]
build_jenkins

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 8]
```diff
@@ -395,43 +395,57 @@ void regclass_graph_Model(py::module m) {
...
+                self.reshape(new_shape, new_variables_shapes);
+            }
         },
         py::arg("partial_shape"),
         py::arg("variables_shapes") = py::dict(),
         R"(
                 Reshape model input.
 
-                The allowed types of keys in the `variables_shapes` dictionary is `str`.
-                The allowed types of values in the `variables_shapes` are:
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:409]: Please, return back these docstrings

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 9]
```diff
@@ -395,43 +395,57 @@ void regclass_graph_Model(py::module m) {
...
+                GIL is released while running this function.
 
-                (1) `openvino.runtime.PartialShape`
-                (2) `list` consisting of dimensions
-                (3) `tuple` consisting of dimensions
-                (4) `str`, string representation of `openvino.runtime.PartialShape`
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:414]: Please, return back these docstrings

**Comment by akuporos** [comment]
Please, return back deleted dosctrings.
Also please resolve conflict

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 10]
```diff
@@ -395,43 +395,57 @@ void regclass_graph_Model(py::module m) {
...
+                        shapes_map[self.input(i)] = partial_shape[i].cast<ov::PartialShape>();
+                    } else if (py::isinstance<py::list>(partial_shape[i]) ||
+                               py::isinstance<py::tuple>(partial_shape[i])) {
+                        shapes_map[self.input(i)] = Common::partial_shape_from_list(partial_shape[i].cast<py::list>());
+                    } else if (py::isinstance<py::str>(partial_shape[i])) {
+                        shapes_map[self.input(i)] = ov::PartialShape(partial_shape[i].cast<std::string>());
+                    }
+                }
+
+                const auto new_variables_shapes = get_variables_shapes(variables_shapes);
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:417]: this line may be moved outside if-else

**Comment by akuporos** [comment]
Hi @amkarn258,

Is there any chance that you will find a time to finish your PR?

**Pull request closed without merging**